File: Challenges/450_DSA_Cracker/013_Kadanes_Algorithm.py
#User function Template for python3

class Solution:
    ##Complete this function
    #Function to find the sum of contiguous subarray with maximum sum.
    def maxSubArraySum(self,arr,N):
        ##Your code here
        
        sumHere = 0
        maxSum = 0
        maxEle = -10E7
        
        for i in range(N) :
            maxEle = max(maxEle, arr[i])
            sumHere += arr[i]
            maxSum = max(maxSum, sumHere)
            sumHere = max(0, sumHere)
        
        if(maxEle < 0) :
            return maxEle
        else : 
            return maxSum

#{ 
 # Driver Code Starts
#Initial Template for Python 3

import math

 
def main():
        T=int(input())
        while(T>0):
            
            n=int(input())
            
            arr=[int(x) for x in input().strip().split()]
            
            ob=Solution()
            
            print(ob.maxSubArraySum(arr,n))
            
            T-=1


if __name__ == "__main__":
    main()
# } Driver Code Ends

File: Challenges/450_DSA_Cracker/Array/013_Kadanes_Algorithm.java

//{ Driver Code Starts
import java.io.*;

import java.util.*;

class Main {
    
	public static void main (String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int t = Integer.parseInt(br.readLine().trim()); //Inputting the testcases
		while(t-->0){
		    //size of array
		    int n = Integer.parseInt(br.readLine().trim());
		    int arr[] = new int[n];
		    String inputLine[] = br.readLine().trim().split(" ");
		    
		    //adding elements
		    for(int i=0; i<n; i++){
		        arr[i] = Integer.parseInt(inputLine[i]);
		    }
		    
		    Solution obj = new Solution();
		    
		    //calling maxSubarraySum() function
		    System.out.println(obj.maxSubarraySum(arr, n));
		}
	}
}


// } Driver Code Ends


class Solution{

    // arr: input array
    // n: size of array
    //Function to find the sum of contiguous subarray with maximum sum.
    long maxSubarraySum(int arr[], int n){
        
        // Your code here
        long maxEle = -10000000;
        long maxSum = 0, sumHere = 0;
        
        for(int i=0; i<n; i++) {
            sumHere += arr[i];
            maxSum = Math.max(maxSum, sumHere);
            sumHere = Math.max(0, sumHere);
            maxEle = Math.max(maxEle, arr[i]);
        }
        
        if(maxEle < 0) {
            return maxEle;
        } else {
            return maxSum;
        }
    }
    
}

